src/IO/H5/VolumeData.cpp

```diff
@@ -54,66 +109,92 @@ VolumeData::VolumeData(const bool exists, detail::OpenGroup&& group,
+           "The expected format of the tensor component names is "
+           "'GROUP_NAME/COMPONENT_NAME' but could not find a '/' in '"
+               << component.name << "'.");
+    return component.name.substr(component.name.find_last_of('/') + 1);
+  };
+  const std::vector<std::string> component_names(
+      boost::make_transform_iterator(elements.front().tensor_components.begin(),
+                                     get_component_name),
+      boost::make_transform_iterator(elements.front().tensor_components.end(),
+                                     get_component_name));
 
-  // Write extents.
-  const auto& extents = extents_and_tensors.extents;
-  if (not contains_attribute(spatial_group.id(), "", "extents")) {
-    h5::write_to_attribute(spatial_group.id(), "extents", extents);
+  // The dimension of the grid is the number of extents per element.
+  // Only written once per VolumeData file, as if two observation id's
+  if (not contains_attribute(volume_data_group_.id(), "", "dimension")) {
+    h5::write_to_attribute(volume_data_group_.id(), "dimension",
+                           elements.front().extents.size());
   }
-
-  // Write the connectivity.
-  if (not h5::contains_dataset_or_group(spatial_group.id(), "",
-                                        "connectivity")) {
-    const std::vector<int> connectivity = [&extents]() noexcept {
-      std::vector<int> local_connectivity;
-      for (const auto& cell : vis::detail::compute_cells(extents)) {
-        for (const auto& bounding_indices : cell.bounding_indices) {
-          local_connectivity.emplace_back(bounding_indices);
-        }
+  size_t dim =
+      h5::read_value_attribute<size_t>(volume_data_group_.id(), "dimension");
+  // Extract Tensor Data one component at a time
+  std::vector<size_t> total_extents;
+  std::string grid_names;
+  std::vector<int> total_connectivity;
+  // We need to keep track the total number of points inserted into the
+  // connectivity after each iteration to be sure each point gets a
+  // unique representation in the topology data
+  int total_points_so_far = 0;
+  // Define functions used to gather element data
+  // Loop over tensor componenents
+  for (size_t i = 0; i < component_names.size(); i++) {
+    std::string component_name = component_names[i];
+    std::vector<double> contiguous_tensor_data{};
+    for (auto& element : elements) {
+      if (i == 0) {  // True if first tensor component being accessed
+        // Factor Out the names and extents
+        append_element_name(&grid_names, element);
+        append_element_extents_and_connectivity(
+            &total_extents, &total_connectivity, &total_points_so_far, dim,
+            element);
       }
-      return local_connectivity;
-    }();
-    h5::write_connectivity(spatial_group.id(), connectivity);
-  }
+      auto tensor_component = element.tensor_components[i];
+      const DataVector& tensor_data_on_grid = tensor_component.data;
```

Comment:
`element.tensor_components[i].data`
and remove
`auto tensor_component = element.tensor_components[i];`